src/VisualStudio/Core/Def/Implementation/Snippets/AbstractSnippetExpansionClient.cs

```diff
@@ -393,7 +451,338 @@ public virtual bool TryInsertExpansion(int startPositionInSubjectBuffer, int end
+                                snippetNamespace + "Header",
+                                new XElement(
+                                    snippetNamespace + "SnippetTypes",
+                                    new XElement(snippetNamespace + "SnippetType", new XText("Expansion"))),
+                                new XElement(snippetNamespace + "Title", new XText(methodName)),
+                                new XElement(snippetNamespace + "Author", "Microsoft"),
+                                new XElement(snippetNamespace + "Description"),
+                                new XElement(snippetNamespace + "HelpUrl"),
+                                new XElement(snippetNamespace + "Shortcut", methodName)),
+                            new XElement(
+                                snippetNamespace + "Snippet",
+                                new XElement(
+                                    snippetNamespace + "Declarations",
+                                    new XElement(
+                                        snippetNamespace + "Literal",
+                                        new XAttribute(snippetNamespace + "Editable", "true"),
+                                        new XElement(snippetNamespace + "ID", new XText("placeholder")),
+                                        new XElement(snippetNamespace + "ToolTip", new XText("")),
+                                        new XElement(snippetNamespace + "Default", new XText("")),
+                                        new XElement(snippetNamespace + "Function"))),
+                                new XElement(
+                                    snippetNamespace + "Code",
+                                    new XAttribute(snippetNamespace + "Language", "csharp"),
+                                    new XCData(template))))));
+
+                var doc = new DOMDocumentClass();
+                if (doc.loadXML(snippet.ToString(SaveOptions.OmitDuplicateNamespaces)))
+                {
+                    var methodSymbols = symbols.OfType<IMethodSymbol>().ToImmutableArray();
+                    _symbols = methodSymbols;
+                    _symbol = null;
+
+                    if (expansion.InsertSpecificExpansion(doc, textSpan, this, LanguageServiceGuid, pszRelativePath: null, out ExpansionSession) == VSConstants.S_OK)
+                    {
+                        Debug.Assert(_symbols == methodSymbols);
+                        Debug.Assert(_symbol == null);
+
+                        // Trigger signature help after starting the snippet session
+                        var workspace = (VisualStudioWorkspace)SubjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges()!.Project.Solution.Workspace;
+                        var exportProvider = (IMefHostExportProvider)workspace.Services.HostServices;
+                        var editorCommandHandlerServiceFactory = exportProvider.GetExports<IEditorCommandHandlerServiceFactory>().Single().Value;
```

Comment:
so this feels really strange.  we don't generally go back to the VSWorkspace to access it's mef provider to pull this stuff out.  instead, we just use MEF to import this stuff in the first place.  Is there a reason the instance of this type can't just be passed in through this type's constructor?

Reply:
IIRC, the limiting factor is unit tests aren't using `VisualStudioWorkspace`

Reply:
➡️ Tracking this under https://github.com/dotnet/roslyn/pull/49916#discussion_r544818461